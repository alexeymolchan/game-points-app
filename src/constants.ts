import { Item } from "./types";

export const itemsMap = {
  "1": {
    id: "1",
    item: "A",
    unitPoints: 50,
    bonus: {
      count: 3,
      points: 200,
    },
  },
  "2": {
    id: "2",
    item: "B",
    unitPoints: 30,
    bonus: {
      count: 2,
      points: 90,
    },
  },
  "3": {
    id: "3",
    item: "C",
    unitPoints: 20,
    bonus: null,
  },
  "4": {
    id: "4",
    item: "D",
    unitPoints: 15,
    bonus: null,
  },
} satisfies Record<string, Item>;

export type ItemId = keyof typeof itemsMap;

export const itemsList = Object.keys(itemsMap) as ItemId[];
