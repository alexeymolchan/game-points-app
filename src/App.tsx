import { useState, MouseEvent, FC } from "react";
import Section from "./components/Section";
import ItemsList from "./components/ItemsList";
import PlayerItemsTable from "./components/PlayerItemsTable";
import Summary from "./components/Summary";
import {
  getItemScoreAndBonusFromQuantity,
  getTotalScoreAndBonuses,
} from "./utils";
import { type ItemId } from "./constants";
import { ItemsMap, PlayerItem } from "./types";
import styles from "./App.module.css";

const INITIAL_STATE = {};

type AppProps = {
  itemsConfig: ItemsMap;
};

const App: FC<AppProps> = ({ itemsConfig }) => {
  const [playerItems, setPlayerItems] =
    useState<Record<string, PlayerItem>>(INITIAL_STATE);

  const resetPoints = () => setPlayerItems(INITIAL_STATE);

  const handleItemClick = (event: MouseEvent<HTMLButtonElement>) => {
    const id = event.currentTarget.dataset.id as ItemId;
    setPlayerItems((prev) => {
      const nextQuantity = (prev[id]?.quantity || 0) + 1;
      const { bonus, unitPoints, name } = itemsConfig[id];
      const { score, bonusPoints } = getItemScoreAndBonusFromQuantity(
        nextQuantity,
        unitPoints,
        bonus
      );

      return {
        ...prev,
        [id]: {
          id,
          quantity: nextQuantity,
          score,
          bonusPoints,
          name,
        },
      };
    });
  };

  const playerItemsData = Object.values(playerItems);

  return (
    <main className={styles.root}>
      <Section className={styles.points} title="Kahoot! Points">
        <div className={styles.container}>
          <h3 className={styles.label}>Items</h3>
          <ItemsList
            items={Object.values(itemsConfig)}
            onItemClick={handleItemClick}
          />
        </div>
      </Section>
      <Section className={styles.playerItems} title="Player Items">
        <PlayerItemsTable data={Object.values(playerItems)} />
        <Summary
          {...getTotalScoreAndBonuses(playerItemsData)}
          onButtonClick={resetPoints}
        />
      </Section>
    </main>
  );
};

export default App;
