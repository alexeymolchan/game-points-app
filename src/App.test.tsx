import { fireEvent, render, screen } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import { itemsMap } from "./constants";
import App from "./App";

const itemsList = Object.values(itemsMap);
const initialBonusesText = "Bonuses 0";
const initialTotalText = "Total 0";

describe("App", () => {
  it("renders initial layout", () => {
    render(<App itemsConfig={itemsMap} />);

    // items
    itemsList.forEach(({ id, name }) => {
      const item = screen.getByTestId(`item-button-${id}`);
      expect(item).toBeInTheDocument();
      expect(item).toHaveTextContent(name);
    });

    // player items
    expect(screen.getByTestId("player-items-root")).toBeInTheDocument();
    expect(screen.getByTestId("player-items-header")).toBeInTheDocument();
    const tableBody = screen.getByTestId("player-items-body");
    expect(tableBody).toBeInTheDocument();
    expect(tableBody.children.length).toBe(0);

    // summary
    expect(screen.getByTestId("summary-bonuses")).toHaveTextContent(
      initialBonusesText
    );
    expect(screen.getByTestId("summary-total")).toHaveTextContent(
      initialTotalText
    );
    expect(screen.getByTestId("new-game-button")).toBeInTheDocument();
  });

  it("add items, calculates total and bonuses", () => {
    render(<App itemsConfig={itemsMap} />);

    const { id: aItemId } = itemsList[0];
    const { id: bItemId } = itemsList[1];

    const bonuses = screen.getByTestId("summary-bonuses");
    const total = screen.getByTestId("summary-total");
    const aItem = screen.getByTestId(`item-button-${aItemId}`);
    const bItem = screen.getByTestId(`item-button-${bItemId}`);

    expect(bonuses).toHaveTextContent(initialBonusesText);
    expect(total).toHaveTextContent(initialTotalText);

    fireEvent.click(bItem);

    expect(
      screen.getByTestId(`player-items-row-${bItemId}`)
    ).toBeInTheDocument();
    expect(bonuses.textContent).toMatchSnapshot();
    expect(total.textContent).toMatchSnapshot();

    fireEvent.click(aItem);

    expect(
      screen.getByTestId(`player-items-row-${aItemId}`)
    ).toBeInTheDocument();
    expect(bonuses.textContent).toMatchSnapshot();
    expect(total.textContent).toMatchSnapshot();

    fireEvent.click(bItem);

    expect(bonuses.textContent).toMatchSnapshot();
    expect(total.textContent).toMatchSnapshot();

    fireEvent.click(aItem);
    fireEvent.click(aItem);

    expect(bonuses.textContent).toMatchSnapshot();
    expect(total.textContent).toMatchSnapshot();
  });

  it("should successfully reset player points", () => {
    render(<App itemsConfig={itemsMap} />);

    const { id: aItemId } = itemsList[0];

    const bonuses = screen.getByTestId("summary-bonuses");
    const total = screen.getByTestId("summary-total");
    const aItem = screen.getByTestId(`item-button-${aItemId}`);
    const newGameButton = screen.getByTestId("new-game-button");

    expect(bonuses).toHaveTextContent(initialBonusesText);
    expect(total).toHaveTextContent(initialTotalText);

    fireEvent.click(aItem);
    fireEvent.click(aItem);
    fireEvent.click(aItem);

    const row = screen.getByTestId(`player-items-row-${aItemId}`);

    expect(row).toBeInTheDocument();
    expect(bonuses.textContent).toMatchSnapshot();
    expect(total.textContent).toMatchSnapshot();

    fireEvent.click(newGameButton);

    expect(screen.getByTestId("player-items-body").children.length).toBe(0);
    expect(row).not.toBeInTheDocument();
    expect(bonuses).toHaveTextContent(initialBonusesText);
    expect(total).toHaveTextContent(initialTotalText);
  });
});
