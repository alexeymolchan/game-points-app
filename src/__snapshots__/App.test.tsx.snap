// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > add items, calculates total and bonuses 1`] = `"Bonuses 0"`;

exports[`App > add items, calculates total and bonuses 2`] = `"Total 30"`;

exports[`App > add items, calculates total and bonuses 3`] = `"Bonuses 0"`;

exports[`App > add items, calculates total and bonuses 4`] = `"Total 80"`;

exports[`App > add items, calculates total and bonuses 5`] = `"Bonuses 30"`;

exports[`App > add items, calculates total and bonuses 6`] = `"Total 140"`;

exports[`App > add items, calculates total and bonuses 7`] = `"Bonuses 80"`;

exports[`App > add items, calculates total and bonuses 8`] = `"Total 290"`;

exports[`App > should successfully reset player points 1`] = `"Bonuses 50"`;

exports[`App > should successfully reset player points 2`] = `"Total 200"`;
